geopandas/clip.py

```diff
@@ -0,0 +1,229 @@
+    """
+    # Create a single polygon object for clipping
+    poly = clip_obj.geometry.unary_union
+    spatial_index = shp.sindex
+
+    # Create a box for the initial intersection
+    bbox = poly.bounds
+    # Get a list of id's for each object that overlaps the bounding box and
+    # subset the data to just those lines
+    sidx = list(spatial_index.intersection(bbox))
+    shp_sub = shp.iloc[sidx]
+
+    # Clip the data - with these data
+    clipped = shp_sub.copy()
+    clipped["geometry"] = shp_sub.intersection(poly)
+
+    # Return the clipped layer with no null geometry values
+    return clipped[clipped.geometry.notnull()]
+
+
+def _clip_multi_poly_line(shp, clip_obj):
+    """Clip multi lines and polygons to the clip_obj GeoDataFrame extent.
+
+    Clip an input multi line or polygon to the polygon extent of the clip_obj
+    parameter. Lines or Polygons that intersect the clip_obj geometry are
+    extracted with associated attributes and returned.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+        multiLine or multipolygon geometry that is clipped to clip_obj.
+
+    clip_obj : GeoDataFrame
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a clipped subset of shp
+        that intersects with clip_obj.
+    """
+
+    # Clip multi polygons
+    clipped = _clip_line_poly(shp.explode().reset_index(level=[1]), clip_obj)
+
+    lines = clipped[
+        (clipped.geometry.type == "MultiLineString")
+        | (clipped.geometry.type == "LineString")
+    ]
+    line_diss = lines.dissolve(by=[lines.index]).drop(columns="level_1")
+
+    polys = clipped[clipped.geometry.type == "Polygon"]
+    poly_diss = polys.dissolve(by=[polys.index]).drop(columns="level_1")
+
+    return gpd.GeoDataFrame(
+        pd.concat([poly_diss, line_diss], ignore_index=True)
+    )
+
+
+def clip_shp(shp, clip_obj):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS) and will
+    be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj,
+    data from shp will be clipped to the total boundary of
+    all polygons in clip_obj.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+          Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame
+          Polygon vector layer used to clip shp.
+          The clip_obj's geometry is dissolved into one geometric feature
+          and intersected with shp.
+
+    Returns
+    -------
+    GeoDataFrame
+         Vector data (points, lines, polygons) from shp clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clipping points (glacier locations in the state of Colorado) with
+    a polygon (the boundary of Rocky Mountain National Park):
+
+        >>> import geopandas as gpd
+        >>> import earthpy.clip as cl
```

Comment:
totally!! i actually had questions about our docs because 
we use earthpy data in our docs and thought we should probably switch that out. Does geopandas have baked in data that you could suggest that i use for this example? Or would you prefer us to create some geometry manually?

i have been contemplating what to do with our docs after this realization.

Reply:
we have an entire vignette that uses earthpy data that will need to be updated!

Reply:
We have `naturalearth_cities`, `naturalearth_lowres`, and `nybb` datasets, so I would use first two (with some buffers etc) as we do in overlay example (http://geopandas.org/set_operations.html)

Reply:
awesome!! thank you. i'll fix this AND the vignette that i hadn't added to this PR just yet.